ecllibrary/std/DataPatterns/Benford.ecl

```diff
@@ -0,0 +1,314 @@
+        int             digitsFound = 0;
+        const char*     ch = s;
+
+        while (*ch)
+        {
+            if (isdigit(*ch) && (pos > 1 || *ch != '0'))
+            {
+                ++digitsFound;
+                if (digitsFound >= pos)
+                {
+                    foundDigit = *ch - 48;
+                    break;
+                }
+            }
+
+            ++ch;
+        }
+
+        return foundDigit;
+    ENDEMBED;
+
+    // Temp field name we will use to ensure proper ordering of results
+    #UNIQUENAME(idField);
+
+    // One-record dataset containing expected Benford results, per-digit
+    #UNIQUENAME(expectedDS);
+    LOCAL %expectedDS% := PROJECT
+        (
+            %expectedDistribution%(pos = %clampedDigit%),
+            TRANSFORM
+                (
+                    {
+                        UNSIGNED2   %idField%,
+                        RECORDOF(LEFT) - [pos],
+                        DECIMAL7_3  chi_squared,
+                        UNSIGNED8   num_values,
+                        STRING      attribute // Put this at the end for now, because it is variable-length
+                    },
+                    SELF.%idField% := 0,
+                    SELF.chi_squared := 0,
+                    SELF.num_values := COUNT(%workingInFile%),
+                    SELF.attribute := '-- EXPECTED DIGIT ' + (STRING)%minDigit% + ' --',
+                    SELF := LEFT
+                )
+        );
+
+    // This will be used later as a datatype in a function signature
+    #UNIQUENAME(DataRec);
+    LOCAL %DataRec% := RECORDOF(%expectedDS%);
+
+    // Get the internal representation of our working dataset
+    #EXPORTXML(inFileFields, RECORDOF(%workingInFile%));
+
+    // Create a dataset composed of the expectedDS and a row for each
+    // field we will be processing
+    #UNIQUENAME(interimResult);
+    LOCAL %interimResult% := %expectedDS%
+        #UNIQUENAME(recLevel)
+        #SET(recLevel, 0)
+        #UNIQUENAME(fieldNum)
+        #SET(fieldNum, 0)
+        #FOR(inFileFields)
+            #FOR(Field)
+                #IF(%{@isRecord}% = 1 OR %{@isDataset}% = 1)
+                    #SET(recLevel, %recLevel% + 1)
+                #ELSEIF(%{@isEnd}% = 1)
+                    #SET(recLevel, %recLevel% - 1)
+                #ELSEIF(%recLevel% = 0)
+                    #SET(fieldNum, %fieldNum% + 1)
+                    + TABLE
+                        (
+                            %workingInFile%(LENGTH(REGEXREPLACE('[^\\d]', (STRING)%@name%, '')) >= %minDigit%),
+                            {
+                                UNSIGNED2   %idField% := %fieldNum%,
+                                DECIMAL4_1  zero := IF(%clampedDigit% = 1, -1.0, COUNT(GROUP, %FirstDigit%((STRING)%@name%, %clampedDigit%) = 0) / COUNT(GROUP) * 100),
+                                DECIMAL4_1  one := COUNT(GROUP, %FirstDigit%((STRING)%@name%, %clampedDigit%) = 1) / COUNT(GROUP) * 100,
```

Comment:
Seems a shame to call function %FirstDigit% multiple times

Reply:
Agreed.  Is there a syntax where I can avoid that, within a TABLE()?

Reply:
Unfortunately I don't think there is a simple way